Initial implementation (#1)
vtex-apps/session-client: docs/README.md

# Session Client

> App that provides React hooks and GraphQL queries to read and update Session.

## Setup

Add this app as a dependency of your React VTEX IO app:

```bash
vtex add vtex.session-client
```

You can have full TypeScript support running `vtex setup --typings` afterwards.

## Usage

### React Hooks

The recommended way to use the session client is through the React hooks provided.

To read session:

- [`useRenderSession`](#useRenderSession)
- [`useFullSession`](#useFullSession)
- [`useLazyFullSession`](#useLazyFullSession)

To update session:

- [`useUpdateSession`](#useUpdateSession)
- [`useUpdateSessionInline`](#useUpdateSessionInline)

#### `useRenderSession`

> The fastest response to a session value. It uses the session response from render-session. One caveat: the session values are limited to a [set of values](https://github.com/vtex-apps/render-session/blob/master/src/constants.ts). If you need fields that are not in this set, you can use [`useFullSession`](#useFullSession) or [`useLazyFullSession`](#useLazyFullSession).

Reviewer: ```suggestion
The fastest way to access a session value. It uses the session response from render-session. One caveat: the session values are limited to a [set of values](https://github.com/vtex-apps/render-session/blob/master/src/constants.ts). If you need fields that are not in this set, you can use [`useFullSession`](#useFullSession) or [`useLazyFullSession`](#useLazyFullSession).
```

```tsx
import React from 'react'
import { useRenderSession } from 'vtex.session-client'

function MyComponent() {
  const { loading, session, error } = useRenderSession()

  if (loading) {
    return <>Session is loading</>
  }

  if (error) {
    return <>Session has errors</>
  }

  console.log({ session })

  return <>Session is ready</>
}

export default MyComponent
```

#### `useFullSession`

> ⚠️ It's not possible to return the session during Server Side Rendering, since that's a private query.

Runs a GraphQL query on the client side to query the full user session.

Under the hood it's just a wrapper of React Apollo's `useQuery` passing the GraphQL session query. You can [read more about the `useQuery` API](https://www.apollographql.com/docs/react/api/react/hooks/#usequery).

**Usage**

```tsx
import React from 'react'
import { useFullSession } from 'vtex.session-client'

function MyComponent() {
  const { loading, data, error } = useFullSession()

  if (loading) {
    return <>Session is loading</>
  }

  if (error) {
    return <>Session has errors</>
  }

  console.log({ session: data?.session })

  return <>Session is ready</>
}

export default MyComponent
```

> 

Reviewer: ```suggestion
```

> It also accepts a GraphQL variable called `items` which is an array of strings of session items to fetch only certain items of the session.

Reviewer: ```suggestion
It also accepts a GraphQL variable called `items` which is an array of `string`s. These `string`s should match attributes inside of the Session object, and only those attributes will be fetched and returned by `useFullSession`.
```

Example:

```tsx
useFullSession({
  variables: {
    items: ['store.channel', 'store.countryCode']
  }
})
```

#### `useLazyFullSession`

> The same as [`useFullSession`](#useFullSession) but it uses the React Apollo's `useLazyQuery` hook. You can [read more about `useLazyQuery` API](https://www.apollographql.com/docs/react/api/react/hooks/#uselazyquery).

Reviewer: ```suggestion
The same as [`useFullSession`](#useFullSession) but it uses React Apollo's `useLazyQuery` hook. You can [read more about `useLazyQuery` API](https://www.apollographql.com/docs/react/api/react/hooks/#uselazyquery).
```

**Usage**

```tsx
import React from 'react'
import { useUpdateSessionInline } from 'vtex.session-client'

function MyComponent() {
  const [updateSession, updatedSession] = useUpdateSessionInline()

  console.log({ updatedSession })

  return (
    <button
      onClick={() =>
        updateSession({
          variables: {
            fields: { foo: 'bar', baz: 123 },
          },
        })
      }
    >
      Update session
    </button>
  )
}

export default MyComponent
```

It also accepts a GraphQL variable called `items` which is an array of strings of session items to fetch only a subset of fields of the session.

Example:

```tsx
updateSession({
  variables: {
    fields: { foo: 'bar', baz: 123 },
    items: ['store.channel', 'store.countryCode']
  }
})
```

### GraphQL

#### session Query

Get the current user session.

**Usage**

```graphql
query session($items: [String]) {
  session(items: $items) @context(provider: "vtex.session-client") {
    ... on SessionSuccess {
      id
      namespaces
    }
    ... on SessionError {
      type
      message
    }
  }
}
```

> #### updateSesion Mutation

Reviewer 2: Just a typo

```suggestion
#### updateSession Mutation
```
